### DISCUSSION TITLE
`knitr::is_html_output()` alternative when using Julia

### DISCUSSION THREAD

**Discussion by kdschneider** [quarto-dev/quarto-cli#5074] [Q&A]
Hey,

I'm trying out writing a report using Julia instead of R. In R I used e.g. `knitr::is_html_output()` to render my plots differently for web and print (e.g. `plotly::ggplotly()` for web). Is there something similar in Julia/Jupyter?

**Comment by cderv** [comment]
Conditional content (https://quarto.org/docs/authoring/conditional.html) will be the closest where you can select which output to use depending on the format. It is not at the same level as execution will still happen, but the plotly HTML won't be included in the PDF output if you use Div with `when-format: html` around the code chunk part.

**Reply by astoeriko** [reply]
Unfortunately, problems arise with this approach (conditional content) when I use the same label for the content in both output formats—Quarto complains about duplicate labels. I will open a new question with an example.

**Reply by mcanouil** [reply]
Duplicate labels complaint is from `knitr` not Quarto, see <https://bookdown.org/yihui/rmarkdown-cookbook/duplicate-label.html>.

**Reply by astoeriko** [reply]
I am using jupyter (with Python), not knitr, so the complaint cannot come from knitr. If I have time tomorrow, I will look into this once more and provide a small reproducing example.

**Reply by cderv** [reply]
Yes Quarto does not allow duplicate cell labels (we use label in various way as ids, and this would be a problem to have duplicate). 

@mcanouil same in **knitr** for sure, but Quarto does enforce the same. 
https://github.com/quarto-dev/quarto-cli/blob/22b46a6326dbdf81287483fde99633495bbca690/src/core/jupyter/labels.ts#L53-L55

**Comment by andrewpbray** [comment]
I have also struggled to include code to produce a plotly plot in a qmd file that is rendered to both HTML and PDF. I finally got something that worked, so I figured I would share it here. Here's a repo with three different attempts that all use conditional content. The third attempt works.

[plotly-pdf.zip](https://github.com/quarto-dev/quarto-cli/files/12824957/plotly-pdf.zip)

*Attempt A*
A naive approach of wrapping the code cell that makes the plotly plot in a conditional content div to only be visible when format is HTML. There is a subsequent conditional div that is visible when format is PDF that shows a png still of that plotly plot.

*Attempt B*
Based on the error message from Attempt A, I tried adding `prefer-html: true` to the document yaml but now it errors out with no message.

*Attempt C*
Use `htmlwidgets::saveWidget()` to create a self-contained html file of the plot, then `htmltools::tags$iframe` in a code cell to include the html file in the HTML output. This one works!

**Reply by cderv** [reply]
@andrewpbray I am surprised that you need to handle this by yourself. **knitr** should automatically handle htmlwidgets output by doing  a screenshot to insert in non HTML output like PDF with LaTeX. 

Though for this to work, **knitr** will look for **webshot** or **webshot2** R package. If available a screenshot will be done. If none a message will be printed output (which have been improved following unexpected behavior like yours) 
The message you should see if you haven't any of this package installed is now 
````
Error:
! Functions that produce HTML output found in document targeting pdf output.
Please change the output type of this document to HTML.
If your aiming to have some HTML widgets shown in non-HTML format as a screenshot,
please install webshot or webshot2 R package for knitr to do the screenshot.
Alternatively, you can allow HTML output in non-HTML formats
by adding this option to the YAML front-matter of
your quarto file:

  prefer-html: true

Note however that the HTML output will not be visible in non-HTML formats.
```` 

As you can see we have now added
> **If your aiming to have some HTML widgets shown in non-HTML format as a screenshot,
please install webshot or webshot2 R package for knitr to do the screenshot.**

So there would be a Attempt D to make which is 
* install **webshot2** R package
* Render the document to PDF without any format conditional 
````markdown
---
title: Attempt A
format:
  html: default
  pdf: default
---

This is an example of a plotly plot. If you are viewing this as an HTML it should be interactive. If you are viewing it as a pdf, it should be a png of the same plot.

```{r}
#| message: false
library(plotly)
fig <- plot_ly(midwest, x = ~percollege, color = ~state, type = "box")
fig
```



````

Is that working and better for you ?

**Reply by andrewpbray** [reply]
Works perfectly. You've shrunk my source code by 2/3 - thanks @cderv !

**Reply by cderv** [reply]
Awesome ! I am glad that works ! 

Happy to help anytime. And I'll think about how to make that easier. We don't want to make **webshot2** a hard dependency but maybe we should be more clear about the warning when not available... 🤔
